File: app/src/main/java/com/senddrop/ui/MainActivity.java
package com.senddrop.ui;

import android.Manifest;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.OpenableColumns;
import android.util.Log;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;
import com.google.zxing.qrcode.QRCodeWriter;
import com.senddrop.android.App;
import com.senddrop.android.R;  // <-- ЭТОТ ИМПОРТ
import com.senddrop.server.HttpServer;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    
    private static final int REQUEST_CODE_PICK_FILE = 1001;
    private static final int REQUEST_PERMISSIONS = 1002;
    
    private HttpServer server;
    private TextView statusText;
    private TextView ipText;
    private Button startButton;
    private Button uploadButton;
    private Button refreshButton;
    private Button copyLinkButton;
    private ImageView qrImage;
    private ListView fileListView;
    private ArrayAdapter<String> fileAdapter;
    private List<String> fileNames = new ArrayList<>();
    private List<File> fileObjects = new ArrayList<>();
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        
        server = new HttpServer();
        
        statusText = findViewById(R.id.statusText);
        ipText = findViewById(R.id.ipText);
        startButton = findViewById(R.id.startButton);
        uploadButton = findViewById(R.id.uploadButton);
        refreshButton = findViewById(R.id.refreshButton);
        copyLinkButton = findViewById(R.id.copyLinkButton);
        qrImage = findViewById(R.id.qrImage);
        fileListView = findViewById(R.id.fileListView);
        
        fileAdapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, fileNames);
        fileListView.setAdapter(fileAdapter);
        
        fileListView.setOnItemClickListener((parent, view, position, id) -> {
            Toast.makeText(this, "Файл: " + fileObjects.get(position).getName(), Toast.LENGTH_SHORT).show();
        });
        
        fileListView.setOnItemLongClickListener((parent, view, position, id) -> {
            deleteFile(position);
            return true;
        });
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (ContextCompat.checkSelfPermission(this, 
                Manifest.permission.WRITE_EXTERNAL_STORAGE) 
                != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE,
                                 Manifest.permission.READ_EXTERNAL_STORAGE}, 
                    REQUEST_PERMISSIONS);
            }
        }
        
        startButton.setOnClickListener(v -> toggleServer());
        uploadButton.setOnClickListener(v -> selectFile());
        refreshButton.setOnClickListener(v -> refreshFileList());
        copyLinkButton.setOnClickListener(v -> copyLink());
        
        refreshFileList();
    }
    
    private void toggleServer() {
        if (server.isRunning()) {
            server.stopServer();
            startButton.setText("Запустить сервер");
            statusText.setText("🔴 Сервер остановлен");
            statusText.setTextColor(getResources().getColor(android.R.color.holo_red_dark));
            qrImage.setImageBitmap(null);
            copyLinkButton.setEnabled(false);
        } else {
            if (server.startServer()) {
                startButton.setText("Остановить сервер");
                statusText.setText("🟢 Сервер запущен: " + server.getUrl());
                statusText.setTextColor(getResources().getColor(android.R.color.holo_green_dark));
                ipText.setText("IP: " + server.getUrl().replace("http://", "").split(":")[0]);
                generateQR(server.getUrl());
                copyLinkButton.setEnabled(true);
                Toast.makeText(this, "Сервер запущен на " + server.getUrl(), Toast.LENGTH_LONG).show();
            } else {
                Toast.makeText(this, "Не удалось запустить сервер", Toast.LENGTH_SHORT).show();
            }
        }
    }
    
    private void generateQR(String data) {
        try {
            QRCodeWriter writer = new QRCodeWriter();
            BitMatrix bitMatrix = writer.encode(data, BarcodeFormat.QR_CODE, 512, 512);
            int width = bitMatrix.getWidth();
            int height = bitMatrix.getHeight();
            Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565);
            for (int x = 0; x < width; x++) {
                for (int y = 0; y < height; y++) {
                    bitmap.setPixel(x, y, bitMatrix.get(x, y) ? Color.BLACK : Color.WHITE);
                }
            }
            qrImage.setImageBitmap(bitmap);
        } catch (WriterException e) {
            Log.e("QR", "Error generating QR: " + e.getMessage());
        }
    }
    
    private void selectFile() {
        if (!server.isRunning()) {
            Toast.makeText(this, "Сначала запустите сервер", Toast.LENGTH_SHORT).show();
            return;
        }
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("*/*");
        startActivityForResult(Intent.createChooser(intent, "Выберите файл"), REQUEST_CODE_PICK_FILE);
    }
    
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE_PICK_FILE && resultCode == RESULT_OK && data != null) {
            Uri uri = data.getData();
            if (uri != null) {
                String filename = getFileName(uri);
                try {
                    File destFile = new File(App.FILES_DIR, filename);
                    if (!destFile.getParentFile().exists()) {
                        destFile.getParentFile().mkdirs();
                    }
                    try (InputStream is = getContentResolver().openInputStream(uri);
                         OutputStream os = new FileOutputStream(destFile)) {
                        byte[] buffer = new byte[8192];
                        int bytesRead;
                        while ((bytesRead = is.read(buffer)) != -1) {
                            os.write(buffer, 0, bytesRead);
                        }
                    }
                    Toast.makeText(this, "Файл загружен: " + filename, Toast.LENGTH_SHORT).show();
                    refreshFileList();
                } catch (Exception e) {
                    Toast.makeText(this, "Ошибка загрузки: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                }
            }
        }
    }
    
    private String getFileName(Uri uri) {
        String result = null;
        if (uri.getScheme().equals("content")) {
            try (Cursor cursor = getContentResolver().query(uri, null, null, null, null)) {
                if (cursor != null && cursor.moveToFirst()) {
                    int nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                    if (nameIndex != -1) {
                        result = cursor.getString(nameIndex);
                    }
                }
            }
        }
        if (result == null) {
            result = uri.getLastPathSegment();
        }
        return result;
    }
    
    private void refreshFileList() {
        fileNames.clear();
        fileObjects.clear();
        List<File> files = server.getFileList();
        for (File f : files) {
            fileNames.add(f.getName() + " (" + formatSize(f.length()) + ")");
            fileObjects.add(f);
        }
        fileAdapter.notifyDataSetChanged();
    }
    
    private void deleteFile(int position) {
        File file = fileObjects.get(position);
        if (file != null && file.exists() && file.delete()) {
            Toast.makeText(this, "Файл удалён: " + file.getName(), Toast.LENGTH_SHORT).show();
            refreshFileList();
        } else {
            Toast.makeText(this, "Не удалось удалить файл", Toast.LENGTH_SHORT).show();
        }
    }
    
    private void copyLink() {
        if (!server.isRunning()) {
            Toast.makeText(this, "Сервер не запущен", Toast.LENGTH_SHORT).show();
            return;
        }
        String url = server.getUrl();
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("SendDrop Link", url);
        clipboard.setPrimaryClip(clip);
        Toast.makeText(this, "Ссылка скопирована: " + url, Toast.LENGTH_SHORT).show();
    }
    
    private String formatSize(long size) {
        if (size < 1024) return size + " B";
        if (size < 1048576) return String.format("%.1f KB", size / 1024.0);
        if (size < 1073741824) return String.format("%.1f MB", size / 1048576.0);
        return String.format("%.1f GB", size / 1073741824.0);
    }
}

File: app/src/main/java/com/senddrop/ui/MainActivity.kt
package com.senddrop.ui

import android.Manifest
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.OpenableColumns
import android.util.Log
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.senddrop.android.App
import com.senddrop.android.Discovery
import com.senddrop.android.Peer
import com.senddrop.android.R
import com.senddrop.android.Transfer
import java.io.IOException
import java.net.NetworkInterface

class MainActivity : AppCompatActivity(), Discovery.OnPeerListChangedListener, Transfer.OnFileReceiveListener {

    private var discovery: Discovery? = null
    private var transfer: Transfer? = null

    // Инициализируем UI без дурацких null
    private lateinit var peerListView: ListView
    private lateinit var peerAdapter: ArrayAdapter<String>
    private val peerNames = mutableListOf<String>()
    private val peerObjects = mutableListOf<Peer>()

    private lateinit var statusText: TextView
    private lateinit var myIpText: TextView
    private lateinit var refreshBtn: Button
    private lateinit var sendFileBtn: Button

    private var selectedPeer: Peer? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        statusText = findViewById(R.id.statusText)
        myIpText = findViewById(R.id.ipText)
        peerListView = findViewById(R.id.fileListView)
        refreshBtn = findViewById(R.id.refreshButton)
        sendFileBtn = findViewById(R.id.uploadButton)
        val copyLinkBtn = findViewById<Button>(R.id.copyLinkButton)

        peerAdapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, peerNames)
        peerListView.adapter = peerAdapter

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.READ_EXTERNAL_STORAGE
                    ),
                    REQUEST_PERMISSIONS
                )
            }
        }

        discovery = Discovery(Build.MODEL, this).apply { start() }
        transfer = Transfer(App.FILES_DIR, this).apply { startServer() }

        val currentIp = localIpAddress
        myIpText.text = "Мой IP: $currentIp"

        // Нормальные Kotlin-лямбды
        refreshBtn.setOnClickListener {
            Toast.makeText(this, "Поиск устройств...", Toast.LENGTH_SHORT).show()
        }

        sendFileBtn.setOnClickListener {
            if (selectedPeer == null) {
                Toast.makeText(this, "Сначала выберите устройство из списка", Toast.LENGTH_SHORT).show()
                return@setOnClickListener // Теперь метка резолвится
            }
            selectFileToSend()
        }

        copyLinkBtn.setOnClickListener {
            val clipboard = getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText("My IP", currentIp)
            clipboard.setPrimaryClip(clip)
            Toast.makeText(this, "IP скопирован: $currentIp", Toast.LENGTH_SHORT).show()
        }

        peerListView.setOnItemClickListener { _, _, position, _ ->
            if (position < peerObjects.size) {
                val peer = peerObjects[position]
                selectedPeer = peer
                statusText.text = "Выбрано: ${peer.name} (${peer.ip})"
                Toast.makeText(this, "Выбрано устройство: ${peer.name}", Toast.LENGTH_SHORT).show()
            }
        }

        statusText.text = "Ожидание устройств..."
    }

    private fun selectFileToSend() {
        val intent = Intent(Intent.ACTION_GET_CONTENT).apply { type = "*/*" }
        startActivityForResult(Intent.createChooser(intent, "Выберите файл"), REQUEST_CODE_PICK_FILE)
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_PICK_FILE && resultCode == RESULT_OK) {
            val uri = data?.data ?: return
            val peer = selectedPeer ?: return

            try {
                contentResolver.openInputStream(uri)?.use { stream ->
                    val fileData = ByteArray(stream.available())
                    stream.read(fileData)

                    // Жестко гарантируем строку, никаких String?
                    val filename = getFileName(uri) ?: "unknown_file"
                    transfer?.sendFile(peer.ip, filename, fileData)

                    Toast.makeText(this, "Отправка файла $filename на ${peer.name}", Toast.LENGTH_LONG).show()
                }
            } catch (e: IOException) {
                Toast.makeText(this, "Ошибка чтения файла: ${e.message}", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun getFileName(uri: Uri): String? {
        var result: String? = null
        if (uri.scheme == "content") {
            contentResolver.query(uri, null, null, null, null)?.use { cursor ->
                if (cursor.moveToFirst()) {
                    val nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                    if (nameIndex != -1) {
                        result = cursor.getString(nameIndex)
                    }
                }
            }
        }
        return result ?: uri.lastPathSegment
    }

    // Возвращаем строго String, а не String?
    private val localIpAddress: String
        get() {
            try {
                val interfaces = NetworkInterface.getNetworkInterfaces()
                for (iface in interfaces) {
                    for (addr in iface.inetAddresses) {
                        if (!addr.isLoopbackAddress && !addr.hostAddress.isNullOrEmpty() && !addr.hostAddress!!.contains(':')) {
                            return addr.hostAddress!!
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e("IP", "Error: ${e.message}")
            }
            return "127.0.0.1"
        }

    // Сигнатуры исправлены на нуллабельные для совместимости с Java
    override fun onPeerAdded(peer: Peer?) {
        if (peer == null) return
        runOnUiThread {
            peerObjects.add(peer)
            peerNames.add("${peer.name} (${peer.ip})")
            peerAdapter.notifyDataSetChanged()
            statusText.text = "Найдено устройств: ${peerObjects.size}"
        }
    }

    override fun onPeerRemoved(peer: Peer?) {
        if (peer == null) return
        runOnUiThread {
            val iterator = peerObjects.iterator()
            var index = 0
            while (iterator.hasNext()) {
                val current = iterator.next()
                if (current.ip == peer.ip) {
                    iterator.remove()
                    if (index < peerNames.size) peerNames.removeAt(index)
                    peerAdapter.notifyDataSetChanged()
                    break
                }
                index++
            }

            if (selectedPeer?.ip == peer.ip) {
                selectedPeer = null
                statusText.text = "Устройство отключено"
            }
            statusText.text = "Найдено устройств: ${peerObjects.size}"
        }
    }

    override fun onFileReceived(filename: String?, size: Long) {
        runOnUiThread {
            val safeName = filename ?: "unknown"
            Toast.makeText(this, "Файл получен: $safeName ($size bytes)", Toast.LENGTH_LONG).show()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        discovery?.stop()
        transfer?.stopServer()
    }

    companion object {
        private const val REQUEST_CODE_PICK_FILE = 1001
        private const val REQUEST_PERMISSIONS = 1002
    }
}